ruby/ql/lib/codeql/ruby/ApiGraphs.qll

```diff
@@ -174,8 +174,7 @@ module API {
           // avoid producing strings longer than 1MB
           result.length() < 1000 * 1000
         )
-      ) and
-      length in [1 .. Impl::distanceFromRoot(this)]
```

Comment:
I don't think we should change this just so we can use it in a test. I think it would be OK to replicate the above in the test itself instead.

Reply:
I removed this because I noticed that we were only calling this predicate in
```ql
string getPath() {
  result = min(string p | p = this.getAPath(Impl::distanceFromRoot(this)) | p)
}
```
which already passes in a max length of `Impl::distanceFromRoot(this)`, so I thought it was redundant. Am I mistaken?

Reply:
>  [...] which already passes in a max length of `Impl::distanceFromRoot(this)`, so I thought it was redundant. Am I mistaken?

Careful! This is "top-down" reasoning ("passes in"), but the evaluation is actually bottom-up. 

Consider how `getAPath` is evaluated in isolation. The lines you deleted would prevent this predicate from containing a `result` and `length` where `length` is greater than the shortest path to the given node. Without those lines, there's no such restriction (except the fact that `result` can't be more than 1000000 characters, which is likely something that happens much further out). In particular, if you have a loop in the API graph (which is almost certain), this predicate will keep going round and round in that loop, producing path strings of greater and greater length, until finally hitting the 1M limit.

In general, it's useful to consider the question "which tuples will be in this relation" _regardless_ of any context that may limit that set. If we're lucky, that context will be automatically magicked in, but it might not (and so it's better to just limit the number of tuples in advance).

Reply:
Thanks for explaining! I'm definitely still getting used to thinking in the bottom-up evaluation mindset :)

Reply:
I've reverted this change in a followup commit - the test now uses a copy of `getAPath` without the length restriction.